lightning/src/util/persist.rs

```diff
@@ -64,41 +114,800 @@ impl<'a, A: KVStorePersister, M: Deref, T: Deref, ES: Deref, NS: Deref, SP: Dere
+	}
+}
+
+/// A struct representing a name for an update.
+#[derive(Clone, Debug, PartialEq, PartialOrd, Eq, Ord)]
+pub struct UpdateName(String);
+
+impl UpdateName {
+	/// The key to store this update with.
+	fn storage_key(&self) -> &str {
+		&self.0
+	}
+}
+
+impl From<u64> for UpdateName {
+	fn from(value: u64) -> Self {
+		Self(format!("{:0>20}", value))
+	}
+}
+
+struct KVStoreUpdatingPersister<K: Deref, L: Deref>
+where
+	K::Target: KVStore,
+	L::Target: Logger,
+{
+	kv: K,
+	logger: L,
+}
+
+impl<K: Deref, L: Deref> KVStoreUpdatingPersister<K, L>
+where
+	K::Target: KVStore,
+	L::Target: Logger,
+{
+	pub fn read_channelmonitors<ES: Deref + Clone, SP: Deref + Clone, B: Deref, F: Deref + Clone>(
+		&self,
+		entropy_source: ES,
+		signer_provider: SP,
+		broadcaster: &B,
+		fee_estimator: F,
+	) -> std::io::Result<
+		Vec<(
+			BlockHash,
+			ChannelMonitor<<SP::Target as SignerProvider>::Signer>,
+		)>,
+	>
+	where
+		ES::Target: EntropySource + Sized,
+		SP::Target: SignerProvider + Sized,
+		B::Target: BroadcasterInterface,
+		F::Target: FeeEstimator,
+	{
+		let mut res = Vec::new();
+		// for each monitor...
+		for monitor_name in self.list_monitor_names()? {
+			// ...parse the monitor
+			let (bh, monitor) = self.deserialize_monitor(
+				entropy_source.clone(),
+				signer_provider.clone(),
+				&monitor_name,
+			)?;
+			// ...parse and apply the updates with an id higher than the monitor.
+			for update_name in self.list_update_names(&monitor_name)? {
+				let update = self.deserialize_monitor_update(&monitor_name, &update_name)?;
+				if update.update_id == CLOSED_CHANNEL_UPDATE_ID
+					|| update.update_id > monitor.get_latest_update_id()
+				{
+					monitor
+						.update_monitor(&update, broadcaster, fee_estimator.clone(), &self.logger)
+						.map_err(|_| KVStoreUpdatingPersisterError::UpdateFailed {
+							reason: "update_monitor returned Err(())",
+							context: &monitor_name.0,
+						})?;
+				}
+			}
+			// ...push the result into the return vec
+			res.push((bh, monitor))
+		}
+		Ok(res)
+	}
+
+	/// Path to corresponding update directory for a given monitor name.
+	fn monitor_update_namespace(&self, monitor_name: &MonitorName) -> String {
+		// Append the monitor name to the namespace with an underscore.
+		[
+			CHANNEL_MONITOR_UPDATE_PERSISTENCE_NAMESPACE,
+			monitor_name.storage_key(),
+		]
+		.join("_")
+	}
+
+	/// List all the names of monitors.
+	fn list_monitor_names(&self) -> io::Result<Vec<MonitorName>> {
+		let list = self.kv.list(CHANNEL_MONITOR_PERSISTENCE_NAMESPACE)?;
+		Ok(list.into_iter().map(MonitorName).collect())
+	}
+
+	/// List all the names of updates corresponding to a given monitor name.
+	fn list_update_names(&self, monitor_name: &MonitorName) -> io::Result<Vec<UpdateName>> {
+		let update_ns = self.monitor_update_namespace(monitor_name);
+		let mut list = self.kv.list(&update_ns)?;
+		list.sort();
+		Ok(list.into_iter().map(UpdateName).collect())
```

Comment:
While I was under the hood doing validation on the newtypes, I also changed this so that update names are first parsed into `u64` and then sorted by their integer values. 

Relying on the left-padding to coincide lexicographic sort with numerical sort was okay in the first iteration I did of this, where I only cared about filesystem. But for the `KVStore` general trait, I think it's unwise to depend on keys being listed the way they were inserted by all backends. For example, I can imagine a relational database being configured to store this as an integer type (maybe incorrectly, but with good intentions). And, since `KVStore.list()` doesn't specify that the keys must come back in lexicographical order, I have to sort anyway -- might as well be by `u64`.

I'm still padding the zeros when representing this as a string, as that would be preferred string-keyed storage (since the natural sort order is correct and will look right when listing ad hoc, etc.).

Reply:
That makes sense. I couldn't trust that lexicographic order and numeric order would always coincide anyway. In practice I think most backends will preserve keys precisely and lexicographically sort when listing, so this is probably often a "sort of a sorted list" [which should run in linear time](https://en.wikipedia.org/wiki/Timsort#Analysis) over a tiny N anyway.